### ISSUE TITLE
storage: START OFFSET should error if larger than partition count

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#23482]
### What version of Materialize are you using?

071643949a

### What is the issue?

Seen in https://buildkite.com/materialize/nightlies/builds/5392#018c120c-8209-496c-b8a1-de5b04c5da8d
```
--- kafka-correctness.td
^^^ +++
kafka-compression.td:70:1: error: non-matching rows: expected:
[["world"]]
got:
[["hello"], ["world"]]
Poor diff:
+ hello

     |
  69 |   FORMAT TEXT
  70 | > SELECT text FROM zstd_fast_forwarded
     | ^
```
This is a regression and reproducible: `bin/mzcompose --find testdrive run default --kafka-default-partitions 5 kafka-compression.td`
I'll bisect it now.

**Comment by def-** [comment]
Actually this code/test was just changed in https://github.com/MaterializeInc/materialize/pull/23436, the bisect also led me to that. Worked fine before it (with `bin/mzcompose --find testdrive run default --kafka-default-partitions 5 zstd-kafka-compression.td` of course)
The start offset here doesn't seem to work:
```sql
> CREATE SOURCE zstd_fast_forwarded
  FROM KAFKA CONNECTION kafka_conn (START OFFSET=[1], TOPIC 'testdrive-zstd-${testdrive.seed}')
  FORMAT TEXT
```
It only fails in the testdrive test with 5 kafka partitions.

**Comment by def-** [comment]
I'm wondering if this is expected behavior? The old test was using 
```
$ kafka-create-topic topic=zstd compression=zstd partitions=1
```
 and now we do
 ```
$ kafka-create-topic topic=zstd compression=zstd
 ```
So apparently Kafka's start offset doesn't work with zstd compression and partitions.

**Comment by philip-stoev** [comment]
If you have multiple partitions, you are supposed to specify an entire list of starting offsets, one per partition. I think the bug here is that the source should have errored out as soon as more partitions than starting offsets were detected.

**Comment by benesch** [comment]
Oh, I actually think this is entirely in spec.

<img width="775" alt="image" src="https://github.com/MaterializeInc/materialize/assets/882976/d8ac4db4-cf7c-48cb-b4b5-a53e9ee769b3">

**Comment by benesch** [comment]
(But the spec declares itself to be incorrect.) @guswynn kindly picked up the fix for this in #23506 (thank you!). Sorry for breaking this, @def-! I didn't realize that we had changed testdrive to allow configuring the default number of partitions for `kafka-create-topic`.
